ocaml/typing/ctype.ml

```diff
@@ -2130,7 +2130,14 @@ let is_immediate64 env ty =
     Btype.backtrack snap;
     result
   else
-    perform_check ()
+    (* CR layouts v2.8: Remove the backtracking once mode crossing is
+       implemented correctly; it's needed for now because checking whether
+       a jkind is immediate (rightly) sets the sort to be Value. It worked
+       previous to this patch because the subjkind check failed earlier. *)
```

Comment:
(I haven't reviewed in detail yet - just a question about this first.)

Can you say more about what the "correct" implementation of mode crossing check will look like?

I think a thing that was true before this PR is that a failing jkind check never changed jkind state.  It could admittedly change some other state, necessitating the backtracking in the previous case.  But the fact about jkinds does still seem like a nice property to have for reasoning about the code.  You say here that it's "right" for a failing check to set the sort variable, but that's not obvious to me - I could imagine it either way, and the other way seems to give that nice reasoning principle.  Probably I'm missing something though?

You mention that it would make the code in `Jkind` uglier to avoid this.  If I understand correctly, it's just a matter of making `Sub_result.combine` take an list of closures and checking them in the given order?  Probably we'll want to generalize `combine` to take a list anyway as jkinds grow, and having that be a list of closures doesn't seem like such a big price to pay.

Reply:
The "correct" implementation will just be a check against externality, not involving the layout at all.

The reason I say it's appropriate for the existing check to set the sort is that the existing check looks to see whether a kind is a subtype of immediate. Following the general design here, every check tries to mutate variables to make the check succeed. The only way for a kind to be a subtype of immediate is for the sort to be `value`. So when we're checking to see whether a kind is a subtype of immediate, we'll set the sort to be `value` if it's a sort variable. The higher-level point is that we shouldn't even be making that check if we don't want to set the sort variable.

As for the failing check not mutating things: that's not true elsewhere. If I try to unify `'a * int` with `char * bool`, I'll link `'a` with `char` before failing. And I think that's just fine. The situation here is very similar.

Reply:
Your comment about "the general design here" isn't obvious to me.  Until now, failing checks didn't mutate sort variables.  So, rather than saying every layout check tries to mutate sort variables to make the check succeed, you could equally well say every check only mutates variables if it will make the check succeed.  And the latter seems nice, even if it's not true for other similar checks in the typechecker.  So the word "rightly" surprised me, because I think either way would work fine.

But, anyway, I don't think this matters so much, particularly since the current implementation is temporary.  I'll try to review this in earnest by early next week.

Reply:
FWIW I think the right position on mutations in such checks is that *consistent* state changes should be preserved. `'a * int` mutating to have `'a` linked with `char` is an example of such a change. An example of an inconsistent state change is when you unify `string t` with `float t`, the first thing that does is mutate the `string t` node to be a link to the `float t` node (which is needed to handle recursive types), but once the inner unification has failed it is important to undo that change and leave things in a consistent state.

The reason to leave consistent state changes in place rather than undo them is that it improves error messages. When you try and make two things equal you want the algorithm to make them as equal as they can be, that way when you print the two things, the only differences between them are the ones that can't be rectified.

(I actually think it would be good to change type unification to do more state changes, so that e.g. unifying `'a * int * 'b` with `string * string * string` would set both `'a` and `'b` to `string`).

Reply:
(FWIW, I am pretty convinced by the argument about better errors that mutating here is indeed preferable, as Richard's original comment suggested)

Reply:
@ccasin is correct about the alternative interpretation of the old implementation -- that we only mutated sort variables if the check would eventually succeed. It was not how I saw things (as my comments suggest!), but it's absolutely true. In any case, I think we agree that this particular choice matters little and that it will be cleaned up soon.